Source/Core/Common/NoDiscard.h

```diff
@@ -0,0 +1,20 @@
+// Copyright 2017 Dolphin Emulator Project
```

Comment:
Indeed, all supported platforms have easy access to a C++17 compiler. Requiring C++17 support would be great for cleanups.

> Sid (which I think the buildbot is on) → GCC 7.2.0 and LLVM 4.0.1

The Ubuntu buildbot runs on 16.04, and the Debian one runs stable AFAIK.

@delroth what do you think of this?

Reply:
Not that it's an actively supported platform but we're currently expecting SteamOS 3.0 to only ship with GCC 6 since that's what Stretch ships. We "might" be able to sidestep the issue with a Dolphin Flatpak if all the dependencies are available and cooperate with the sandbox (that's what I've been doing with RetroArch and its prebuilt cores on SteamOS 2.0) but for those who want to build on the box itself, requiring 7+ would throw a wrench into that.

Reply:
Does SteamOS not have a package management system of some sort? I thought it was a variant of Debian (I haven't used Steam OS and I've only read little about it, so correct me if I'm wrong). Would it not be possible to use a different repository source or build GCC 7?

I personally don't particularly agree with holding back the codebase for SteamOS. Especially so if it's a Linux variant that happens to not allow their end-users/end-developers(?) to update their own compilers and related tools if desirable. It's also not what I would necessarily call a main platform for development on. (e.g. most, if not all, core devs for Dolphin use an unrestricted Linux distribution with regards to software). I'm also not particularly comfortable with the idea of essentially being in lockstep with SteamOS's update pattern when it comes to our compiler versions (or, any utility, really). That's just my take on it, anyways.

Reply:
Linux is less than a % of our userbase.

I don't care about SteamOS. And SteamOS is just debian based.

Reply:
@lioncash It's possible to add Debian repositories to the SteamOS package manager but I don't know how easy it is to get g++-7 there:

https://github.com/mdeguzis/SteamOS-Tools/wiki/Desktop-Software

Is that something `debian-backports` would have...? I'm not terribly familiar with Debian so my knowledge on what's backported and what isn't is pretty limited.

I point out SteamOS mostly because it's my use case but from a broader perspective this is more about Debian Stretch compatibility than anything else.

Reply:
There is no easy way to get GCC 7 on Stretch, as stretch-backports doesn't have it. Personally, I wouldn't worry about that too much. Because of how quickly the packages become outdated (for non-server use), it's likely very few people use stable versions of Debian for their desktop...

Reply:
I'm not certain about backports, however installing them from sid individually does work. e.g.

Adding sid to the list of package sources:

```bash
deb http://ftp.us.debian.org/debian/ sid main
deb-src http://ftp.us.debian.org/debian/ sid main
```

and then doing:

```bash
apt-get -t sid install g++
```

should be sufficient on Stretch. In the cases where I need to use Debian, I usually end up doing this for one reason or another, and I've been able to build Dolphin fine with it in the past.

Reply:
@lioncash Spent some time trying this out on a test box and eventually it does work! Sadly it's not _nearly_ as clean to do as it is for vanilla Debian but after slamming apt's head against a wall for a while it's possible to build/run upstream Dolphin without breaking the rest of the system. So there's another route we can take if we need it.

Reply:
@lioncash created a separate PR for C++17. It's likely to fail on all builders before they are updated to run a C++17 compatible compiler, though.